### ISSUE TITLE
Eval bug: NaN logits on CUDA for specific k-quant mixes of a Laguna (118B MoE) model; CPU clean; test-backend-ops MUL_MAT_ID passes

### ISSUE THREAD

**Issue by erm14254** [ggml-org/llama.cpp#27899]
## Summary

Several k-quant variants of a `laguna`-arch model (poolside Laguna-S-2.1 based, 48 layers, 256-expert MoE, 3072 hidden) produce NaN logits when evaluated with the CUDA backend, while the **same GGUF files evaluate cleanly on CPU** (`-dev none`). The failure is content-dependent (plain English wikitext triggers it; a C++ code corpus often does not) and rounding-dependent (requantizing the same recipe with an imatrix moves the trigger point). `test-backend-ops test -b CUDA0 -o MUL_MAT_ID` passes, so the generic unit tests do not reproduce it — real model weights/shapes appear to be required.

## Environment

- llama.cpp master `cb300598d`, CUDA build (MSVC, Windows 11)
- GPU: RTX PRO 6000 Blackwell Workstation Edition 96GB (sm_120); driver-recent
- Repro observed at every offload level tried (`-ngl 99`, `-ngl 0`) and with `-fa on` and `-fa off`; only `-dev none` (pure CPU) is clean

## Repro

```
llama-perplexity -m <model-Q3_K_L>.gguf -f wiki.test.raw -c 512 --chunks 6
# chunks: [1]4.57,[2]6.50,[3]-nan,... then "Unexpected negative standard deviation of log(prob)"

llama-perplexity -m <same file> -f wiki.test.raw -dev none -c 512 --chunks 6
# clean: [1]4.65,[2]6.53,[3]6.17,[4]7.40,[5]7.63,[6]8.68
```

wiki.test.raw = standard wikitext-2-raw test set. The affected chunk is ordinary English text.

## Which quants fail (all built from the same BF16 with identical protective overrides: attention/shared-experts/dense-FFN/embd/output kept BF16; only routed experts quantized)

| Quant | routed-expert types (gate/up + down) | CUDA | CPU |
|---|---|---|---|
| Q2_K | q2_K + q3_K | clean (120ch × 2 corpora) | clean |
| Q3_K_S | q3_K + q3_K | clean (120ch × 2) | clean |
| **Q3_K_M** | q3_K + q4_K(45)/q5_K(2) | **NaN @ wiki chunk 3** | clean |
| **Q3_K_L** | q3_K + q5_K | **NaN @ wiki chunk 3** (imatrix rebuild of same recipe: NaN @ chunk 18) | clean |
| **Q4_K_S** | q4_K + q4_K(42)/q5_K(5) | **NaN** (240-chunk combined corpus) | — |
| Q4_K_M | q4_K + q4_K(24)/q6_K(23) | clean (240ch) | — |
| **Q5_K_S** | q5_K + q5_K (uniform) | **NaN** (240ch) | — |
| Q5_K_M | q5_K + q5_K(24)/q6_K(23) | clean (240ch) | — |
| Q6_K | q6_K uniform | clean (240ch) | clean |
| Q8_0 | q8_0 | clean (240ch) | — |
| BF16 | none | clean (36ch) | — |

Observations:

- Uniform q5_K (Q5_K_S) fails while Q5_K_M — which still has q5_K gate/up on all layers and q5_K down on half — passes. Likewise Q4_K_S fails while Q4_K_M passes. The _M variants differ from _S mainly by upgrading `ffn_down_exps` to q6_K on roughly half the layers.
- Working hypothesis (unproven): one or more specific layers' `ffn_down_exps` tensors, when encoded as q4_K/q5_K, produce extreme values in the CUDA path for particular activations; q3_K/q6_K/q8_0/BF16 encodings of the same tensors are fine, and the _M-variant q6_K upgrades happen to cover the responsible layer(s).
- `test-backend-ops test -b CUDA0 -o MUL_MAT_ID`: all OK.
- Possibly related in spirit: #25455 (SYCL MUL_MAT_ID corruption incl. q3_K on MoE), though that reporter finds CUDA clean in unit tests, as do we.

## Root-cause isolation (eval-callback)

Using `llama-eval-callback` patched to abort on the first non-finite tensor sum (one-line change: enable the existing `abort_on_nan` and extend it to Inf), all three quants tested converge on the **same node**:

```
ffn_moe_swiglu-45 : finite, small values (samples in ±1.6, full-tensor sum ~9.8e4)
ffn_moe_down-45   = MUL_MAT_ID(blk.45.ffn_down_exps.weight{1024,3072,256}, swiglu{1024,10,512})
                  -> output contains ±Inf (sum = -inf)
```

- Q3_K_L (blk.45 down = q5_K): abort at `ffn_moe_down-45`
- Q4_K_S (blk.45 down = q4_K): abort at `ffn_moe_down-45` (different trigger text)
- Q5_K_S (blk.45 down = q5_K): process hard-stops while dumping `ffn_moe_down-45` (no abort message — possibly a crash in the same region)

Finite small inputs × quantized weights cannot legitimately produce Inf at this scale (1024-element dot products), so the CUDA MUL_MAT_ID path appears to compute garbage for this specific tensor's data (bad block/scale read, indexing, or dequant edge case). n_expert = 256, n_expert_used = 10.

## Workaround (validated)

Requantizing with only `blk.45.ffn_down_exps` overridden to q6_K (`--tensor-type 'blk\.45\.ffn_down_exps=q6_k'`, ~200 MiB size cost) makes all four affected presets pass 240-chunk × 2-corpus perplexity runs — including the exact trigger chunks — plus generation smoke tests. An all-layers `ffn_down_exps=q6_k` diagnostic build also passes. This confirms the failure is tied to the q4_K/q5_K encodings of this particular tensor on CUDA.

## Model availability

The model is a public-weights 118B MoE (Laguna-S-2.1 derivative); the failing GGUFs are 62-85GB. I can share the exact failing files, the trigger text, and run any diagnostic patches/builds on the affected hardware.
